### ISSUE TITLE
current working directory not restored after subshell if not searchable

### ISSUE THREAD

**Issue by stephane-chazelas** [ksh93/ksh#141]
```
# pwd
/root
# ls -ld .
drwx------ 14 root root 125 Oct  4 07:46 ./
# id -u
0
# sudo -u nobody ksh -c '(cd /); cd -P .; pwd -P; /bin/pwd'
/root
/
```

The `cd -P .` being a failed attempt to make the builtin `pwd` acknowledge the real current working directory.

That's because subshells in ksh93 are not implemented by forking. To be able to restore the working directory, ksh93 opens `.` on some fd and does a `fchdir(fd)` upon termination of the subshell.

But here, `open(".")` fails as `.` is not searchable.

A fix could be to use a forking subshell when `open(".")` fails (using `ulimit -t unlimited` to force a fork does work around the issue here), but that would not address the case where `.` becomes non-searchable *after* a successful `open(".")`, and where `fchdir(fd)` could fail (as in `(chmod 0 .; cd /)`).

**Comment by McDutchie** [comment]
This problem is inherent to non-forking subshells, isn't it? If you `chdir` out of the present working directory in the current process, there is never a guarantee you can `chdir` back again.

I don't see how this could possibly be fixed reliably except by always forking a subshell upon encountering `cd`. And that would be a performance regression that I am certain the community would find unacceptable.

**Comment by stephane-chazelas** [comment]
The problem is made worse here by the fact that the problem is not reported. Worse, it is hidden (`pwd` makes you think the current working directory has been successfully restored).

It can have dramatic consequences as in:

```
blah; (cd / && pwd); rm -rf -- *
```

I'd suggest:

1. perform a forking subshell when `open(".")` (to save a copy of the cwd) fails
2. write an error message and exit the shell (to limit the damage) when `fchdir()` (to restore the cwd) fails.

Since that's about uncommon pathological condition, that shouldn't have a performance impact.

**Comment by McDutchie** [comment]
> ```
> # ls -ld .
> drwx------ 14 root root 125 Oct  4 07:46 ./
> # id -u
> 0
> # sudo -u nobody ksh -c '(cd /); cd -P .; pwd -P; /bin/pwd'
> /root
> /
> ```
BTW, it behaves differently on macOS, I've no idea why, but it does not report a false PWD:
```
# uname -sr; whoami
Darwin 18.7.0
root
# ls -ld .
drwxr-x---  14 root  wheel  448 Jun 17 19:45 .
# sudo -u nobody /usr/local/src/ksh93/ksh/arch/darwin.i386-64/bin/ksh -c '(cd /); cd -P .; pwd -P; /bin/pwd'
/
/
# sudo -u nobody /bin/ksh -c '(cd /); cd -P .; pwd -P; /bin/pwd'
/
/
```

**Comment by McDutchie** [comment]
> I'd suggest:
> 
> 1. perform a forking subshell when `open(".")` (to save a copy of the cwd) fails
> 2. write an error message and exit the shell (to limit the damage) when `fchdir()` (to restore the cwd) fails.

Well, the second point is easy, so let's do that first…
```diff
diff --git a/src/cmd/ksh93/sh/subshell.c b/src/cmd/ksh93/sh/subshell.c
index 0de8fe9..97d6681 100644
--- a/src/cmd/ksh93/sh/subshell.c
+++ b/src/cmd/ksh93/sh/subshell.c
@@ -759,13 +759,11 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 			Namval_t *pwdnod = sh_scoped(shp,PWDNOD);
 			if(shp->pwd)
 			{
-				if(sp->pwdfd >=0)
-				{
-					if(fchdir(sp->pwdfd)<0)
-						chdir(sp->pwd);
-				}
-				else
-					chdir(sp->pwd);
+				int result;
+				if(sp->pwdfd < 0 || (result = fchdir(sp->pwdfd)) < 0)
+					result = chdir(sp->pwd);
+				if(result < 0)
+					errormsg(SH_DICT,ERROR_system(128),"failed to restore PWD upon exiting subshell");
 				shp->pwd=sp->pwd;
 				path_newdir(shp,shp->pathlist);
 			}
```
The first item is more complicated. If `open(".",O_RDONLY);` fails in: https://github.com/ksh93/ksh/blob/6697edba1c302c4bdd0796945158decc825b5bd5/src/cmd/ksh93/sh/subshell.c#L545-L564 …then we're smack in the middle of creating a virtual subshell, so forking it does not seem possible. Normally, a subshell is forked with [`sh_subfork()`](https://github.com/ksh93/ksh/blob/6697edba1c302c4bdd0796945158decc825b5bd5/src/cmd/ksh93/sh/subshell.c#L168-L215) but that immediately crashes in this context – probably because we just did a [`sh_pushcontext()`](https://github.com/ksh93/ksh/blob/6697edba1c302c4bdd0796945158decc825b5bd5/src/cmd/ksh93/include/fault.h#L104-L109) and haven't popped the context yet; I've noticed `sh_subfork()` always crashes then.

**Comment by McDutchie** [comment]
Good news: `sh_subfork()` turns out to work fine if we simply execute it immediately before the `sh_exec()` that executes the command tree for the subshell. So we can just delay the check.

The following diff fixes it for me. Please test.
```diff
diff --git a/src/cmd/ksh93/sh/subshell.c b/src/cmd/ksh93/sh/subshell.c
index 0de8fe9..8dcf790 100644
--- a/src/cmd/ksh93/sh/subshell.c
+++ b/src/cmd/ksh93/sh/subshell.c
@@ -633,6 +633,8 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 		if(shp->savesig < 0)
 		{
 			shp->savesig = 0;
+			if(sp->pwdfd < 0 && !shp->subshare)	/* if we couldn't get a file descriptor to our PWD ... */
+				sh_subfork();			/* ...fork to avoid ending up in the wrong PWD on exit */
 			sh_offstate(SH_INTERACTIVE);
 			sh_exec(t,flags);
 		}
@@ -759,13 +761,11 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 			Namval_t *pwdnod = sh_scoped(shp,PWDNOD);
 			if(shp->pwd)
 			{
-				if(sp->pwdfd >=0)
-				{
-					if(fchdir(sp->pwdfd)<0)
-						chdir(sp->pwd);
-				}
-				else
-					chdir(sp->pwd);
+				int result;
+				if(sp->pwdfd < 0 || (result = fchdir(sp->pwdfd)) < 0)
+					result = chdir(sp->pwd);
+				if(result < 0)
+					errormsg(SH_DICT,ERROR_system(128),"failed to restore PWD upon exiting subshell");
 				shp->pwd=sp->pwd;
 				path_newdir(shp,shp->pathlist);
 			}
```
Behaviour on macOS, before this diff:
```
$ mkdir bad && cd bad && chmod 0 . && { "$OLDPWD"/arch/*/bin/ksh -xc '(cd /); cd -P .; pwd -P; /bin/pwd'; cd "$OLDPWD"; rmdir bad; }
+ cd /
+ cd -P .
+ pwd -P
/
+ /bin/pwd
/
```
Behaviour on macOS, after this diff:
```
$ mkdir bad && cd bad && chmod 0 . && { "$OLDPWD"/arch/*/bin/ksh -xc '(cd /); cd -P .; pwd -P; /bin/pwd'; cd "$OLDPWD"; rmdir bad; }
+ cd /
+ cd -P .
/usr/local/src/ksh93/ksh/arch/darwin.i386-64/bin/ksh: cd: .: [Permission denied]
+ pwd -P
pwd: cannot access parent directories [Permission denied]
+ /bin/pwd
pwd: .: Permission denied
```

**Comment by McDutchie** [comment]
Thinking aloud some more…

Perhaps the above diff forks prematurely. There is an argument to be made that forking is only necessary when the PWD is invalid/inaccessible *and* we encounter the `cd` builtin. I'm pretty sure that nothing else changes the working directory in the normal ksh language.

Then again, ksh supports dynamically loadable custom builtins and there's no guarantee one of those won't change the directory. Not that I'm aware of any actual use of that feature...

**Comment by McDutchie** [comment]
On restoring the PWD, the original ksh code tries `fchdir`(2) first and if that fails, it tries falling back to `chdir`(2). I kept that fallback in the diff above. But there's no guarantee that the same path still points to the same directory after completing the subshell. For example, an attacker might have moved the directory and put a symlink in its place. Whereas the file descriptor would be immune to renames, and fail appropriately if the directory no longer exists even if a different directory with the same path was created in the meantime. So maybe that `chdir` fallback needs to go.

I notice the `fchdir` use is subject to a feature test, though. Is there actually still any system around that does not have this [POSIX-specified](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fchdir.html) syscall? On systems without `fchdir` (if any), maybe we just need to make `cd` fork unconditionally, so that the parent process never needs to restore the PWD. And if some custom builtin changes the PWD, there is simply nothing we can do in that case.

**Comment by McDutchie** [comment]
@stephane-chazelas, I'd appreciate your critical eyes on my fix (dd9bc22928c71376d14e686770505ec119f2b46f). Thanks for the report in any case!
